openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -408,24 +417,31 @@ def _refresh_addresses(self):
         handle cases in which the IP has changed
         """
         self.connection.device.refresh_from_db()
-        self.connection.refresh_from_db()
+        self.connection = self.connection.get_working_connection(self.connection.device)
         self.addresses = self.connection.get_addresses()
 
     def _write_checksum(self, checksum):
         for attempt in range(1, self.RECONNECT_MAX_RETRIES + 1):
-            self._refresh_addresses()
-            addresses = ', '.join(self.addresses)
-            self.log(
-                _(
-                    'Trying to reconnect to device at {addresses} (attempt n.{attempt})...'.format(
-                        addresses=addresses, attempt=attempt
-                    )
-                ),
-                save=False,
-            )
             try:
+                self._refresh_addresses()
+                addresses = ', '.join(self.addresses)
+                self.log(
+                    _(
+                        'Trying to reconnect to device at {addresses} (attempt n.{attempt})...'.format(
+                            addresses=addresses, attempt=attempt
+                        )
+                    ),
+                    save=False,
+                )
                 self.connect()
```

Comment:
While performing manual testing locally, I got this log for the device 

```
Connection successful, starting upgrade...
Image checksum file not found, proceeding with the upload of the new image...
Sysupgrade test passed successfully, proceeding with the upgrade operation...
Upgrade operation in progress...
Commencing upgrade. Closing all shell sessions.
Image metadata not found
Reading partition table from bootdisk...
Reading partition table from image...
Partition layout has changed. Full image will be written.

SSH connection closed, will wait 150 seconds before attempting to reconnect...
Device not reachable yet, (connection failed).
retrying in 30 seconds...
Device not reachable yet, (connection failed).
retrying in 30 seconds...
Trying to reconnect to device at 192.168.56.2 (attempt n.3)...
Connected! Writing checksum file to /etc/openwisp/firmware_checksum
Upgrade completed successfully.
``` 

See `Trying to reconnect to device at 192.168.56.2 (attempt n.3)...` shows third attempt. This is because in previous two attempts the device was unreachable and `self._refresh_addresses()` raised `NoWorkingDeviceConnectionError` which led to the logs getting skipped. 

How can we get updated addresses without initiating a connection?

Reply:
what about:

```python
```python
def _log(addresses, attempt):
    self.log(
        _(
            f'Trying to reconnect to device at {addresses} '
            '(attempt n.{attempt})...'.format(
                addresses=addresses,
                attempt=attempt
            )
        ),
        save=False,
    )
try:
    self._refresh_addresses()
except NoWorkingDeviceConnectionError as error:
    _log(error.connection.addresses, attempt)
try:
    _log(', '.join(self.addresses), attempt)
    self.connect()
except (NoValidConnectionsError, socket.timeout, SSHException) as error:
    # etc...
```